lib/rust/data-structures/src/counter.rs

```diff
@@ -0,0 +1,98 @@
+//! # Counter
+//!
+//! Counter type. Uses an internal global value to ensure every instance created has a different
+//! value.
+
+use crate::prelude::*;
+
+
+
+// ===============
+// === Counter ===
+// ===============
+
+/// Implements a globally-unique counter.
+#[derive(Clone, CloneRef, Copy, Debug, Eq, Hash, Into, PartialEq, Ord, PartialOrd)]
+pub struct Counter {
+    value: u64,
+}
+
+/// Generate a unique value.
+impl Default for Counter {
```

Comment:
Is it a good idea to do this via default? 
This means that  `Counter::default() != Counter::default()` which seems wrong.

Reply:
Yeah, I personally don't like `default` calls for constructing *different* values.

Reply:
Good catch. I provided an `impl Default` because a lot of code relies on the ID types being default-constructable--but looking at how the types are used, allocating real IDs in `default()` is wrong (in addition to violating expectations of `Default`); the existing IDs are actually nullable types. That stinks. I will change the ID implementation to be based on `Option<Counter>` to support current usages of the types, and I have opened [an issue for refactoring it in the future](https://www.pivotaltracker.com/n/projects/2539304/stories/181626362).

Reply:
Actually, I'm going to refine this: Most users of `define_id` don't need an impl default, so I will move the Option to the one that does currently need nullability.

Reply:
Nevermind that, we do have multiple users of IDs that rely on nullability. I'll leave it as is.

Reply:
@kazcw If we don't implement that properly now, we will never do it properly. So IMO either we should leave current IDs, or we should introduce new IDs and use `Option<Id>` in places where current IDs are nullable. IDs should not be nullable, so making them nullable by definition is IMO a wrong decision. Even if you create an issue, no one will touch the issue for the next months / years. Would it be very time consuming to finish it properly now?

Reply:
There are probably cases where `default()` has the effect of providing a sensible initial value; there are cases (like we're encountering here) where no such sensible default exists. We could distinguish the two cases if we did something like:
```rust
pub trait Data {
    type WithDefault: Default;
}
```

Then for types where Default provides a reasonable initial value, we could `impl Data` with `type WithDefault = Self`; for other types, `type WithDefault = Option<Self>`.
We'd need to write a lot of `impl Data` declarations and change some instances of `T` to `<T as Data>::WithDefault`, but those are straightforward changes, so I think this approach wouldn't require too much work. Everywhere we'd need to change a `T` to an `Option<T>` is a place we should check we're handling null values correctly.

Reply:
FRP is designed this way and I believe it is a correct design, because if you are using a sampler and sampling an event output that was not yet emitted, a default value should be delivered. Anyway, this is a longer discussion if that's correct and if there are better ways of doing it. Anyway, you can still change the code in such a way that instead f passing ID, you can pass Option<Id> there, right?

Reply:
Ok, makes sense. So let's make these IDs nullable for now and lets see how we can improve that in the future. Thanks for super detailed reply and explanation here!

Reply:
I was wrong earlier in this thread when I said there's more than one use of `define_id` that needs `impl Default` (I mixed up the two `struct Id` cases). There is actually only one, so what I can do is: Remove the `Option` from the `define_id` implementation and move it to the id type in `display::object::class`. This means that case won't be able to use the `RcWithId` type, but I think it's OK if the highest-level interface doesn't support doing this thing that ideally we wouldn't be doing.